NervanaSystems/distiller
# ADC (Automatic Deep Compression) example + features, tests, bug fixes

This is a merge of the ADC branch and master.
ADC (using a DDPG RL agent to compress image classifiers) is still WiP and requires 
An unreleased version of Coach (https://github.com/NervanaSystems/coach).
 
Small features in this commit:
---------------------------------------
-Added model_find_module() - find module object given its name
-Added tests:
test_summarygraph.py: test_simplenet() - Added a regression test to target a 
         bug that occurs when taking the predecessor of the first node in a graph
test_ranking.py - test_ch_ranking, test_ranked_channel_pruning
test_model_summary.py - test_png_generation, test_summary (sparsity/
          compute/model/modules)
- add channel ranking and pruning: pruning/ranked_structures_pruner.py
- Add a CIFAR10 VGG16 model: models/cifar10/vgg_cifar.py
- Thinning: change the level of some log messages – some of the messages were
               moved to ‘debug’ level because they are not usually interesting.
- Add a function to print nicely formatted integers - distiller/utils.py
- Sensitivity analysis for channels-removal
- compress_classifier.py – handle keyboard interrupts
- compress_classifier.py – fix re-raise of exceptions, so they maintain call-stack

Bug fixes in this commit:
- Thinning bug fix: handle zero-sized 'indices' tensor
During the thinning process, the 'indices' tensor can become zero-sized,
              and will have an undefiend length.  Therefore, we need to check for this
              situation when assessing the number of elements in 'indices'
- Language model: adjust main.py to new distiller.model_summary API
